dlt/destinations/impl/clickhouse/clickhouse.py

```diff
@@ -0,0 +1,416 @@
+            file_path,
+            table["name"],
+            self.sql_client,
+            staging_credentials=(
+                self.config.staging_config.credentials if self.config.staging_config else None
+            ),
+        )
+
+    def _get_table_update_sql(
+        self, table_name: str, new_columns: Sequence[TColumnSchema], generate_alter: bool
+    ) -> List[str]:
+        table: TTableSchema = self.prepare_load_table(table_name, self.in_staging_mode)
+        sql = SqlJobClientBase._get_table_update_sql(self, table_name, new_columns, generate_alter)
+
+        if generate_alter:
+            return sql
+
+        # Default to 'ReplicatedMergeTree' if user didn't explicitly set a table engine hint.
+        table_type = cast(
+            TTableEngineType, table.get(TABLE_ENGINE_TYPE_HINT, "replicated_merge_tree")
+        )
+        sql[0] = f"{sql[0]}\nENGINE = {TABLE_ENGINE_TYPE_TO_CLICKHOUSE_ATTR.get(table_type)}"
+
+        if primary_key_list := [
+            self.capabilities.escape_identifier(c["name"])
+            for c in new_columns
+            if c.get("primary_key")
+        ]:
+            sql[0] += "\nPRIMARY KEY (" + ", ".join(primary_key_list) + ")"
+        else:
+            sql[0] += "\nPRIMARY KEY tuple()"
+
+        return sql
+
+    def get_storage_table(self, table_name: str) -> Tuple[bool, TTableSchemaColumns]:
+        fields = self._get_storage_table_query_columns()
+        db_params = self.sql_client.make_qualified_table_name(table_name, escape=False).split(
+            ".", 3
+        )
+        query = f'SELECT {",".join(fields)} FROM INFORMATION_SCHEMA.COLUMNS WHERE '
+        if len(db_params) == 3:
+            query += "table_catalog = %s AND "
+        query += "table_schema = %s AND table_name = %s ORDER BY ordinal_position;"
+        rows = self.sql_client.execute_sql(query, *db_params)
+
+        # If no rows we assume that table does not exist.
+        schema_table: TTableSchemaColumns = {}
+        if len(rows) == 0:
+            return False, schema_table
+        for c in rows:
+            numeric_precision = (
+                c[3] if self.capabilities.schema_supports_numeric_precision else None
+            )
+            numeric_scale = c[4] if self.capabilities.schema_supports_numeric_precision else None
+            schema_c: TColumnSchemaBase = {
+                "name": c[0],
+                "nullable": bool(c[2]),
+                **self._from_db_type(c[1], numeric_precision, numeric_scale),
+            }
+            schema_table[c[0]] = schema_c  # type: ignore
+        return True, schema_table
+
+    def get_stored_schema(self) -> StorageSchemaInfo:
+        return super().get_stored_schema()
+
+    @staticmethod
+    def _gen_not_null(v: bool) -> str:
+        # ClickHouse fields are not nullable by default.
+        # We use the `Nullable` modifier instead of NULL / NOT NULL modifiers to cater for ALTER statement.
```

Comment:
OK. but I have a test that adds NON NULL column to existing column with the data and it passes. so we are not adding this constraint or it is impossible?
`test_many_schemas_single_dataset`

Reply:
see above